From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: quh4gko8 <88831734+quh4gko8@users.noreply.github.com>
Date: Sat, 12 Jul 2025 16:30:34 +0000
Subject: [PATCH] set multiArch=false in 64-bit only builds of browser and
 webview

Since the placeholder libraries are not included in 64-bit only builds,
and it is not required for browser and webview apk build targets,
set multiArch=false to allow its installation on OS with 32-bit support.
---
 android_webview/nonembedded/java/AndroidManifest.xml     | 4 ++++
 android_webview/system_webview_apk_tmpl.gni              | 2 ++
 chrome/android/chrome_public_apk_tmpl.gni                | 4 ++++
 chrome/android/java/AndroidManifest_monochrome.xml       | 4 ++++
 chrome/android/java/AndroidManifest_trichrome_chrome.xml | 4 ++++
 5 files changed, 18 insertions(+)

diff --git a/android_webview/nonembedded/java/AndroidManifest.xml b/android_webview/nonembedded/java/AndroidManifest.xml
index 6ac564f40754a..93a4652145485 100644
--- a/android_webview/nonembedded/java/AndroidManifest.xml
+++ b/android_webview/nonembedded/java/AndroidManifest.xml
@@ -43,7 +43,11 @@ by a child template that "extends" this file.
     <application android:label="{{ application_label|default('Vanadium System WebView') }}"
                  android:icon="@{{manifest_package|default('com.android.webview')}}:drawable/icon_webview"
                  android:name="{{ application_name|default('org.chromium.android_webview.nonembedded.WebViewApkApplication') }}"
+                 {% if has_secondary_abi is defined and has_secondary_abi == 'false' %}
+                 android:multiArch="false"
+                 {% else %}
                  android:multiArch="true"
+                 {% endif %}
                  android:memtagMode="async"
                  android:extractNativeLibs="false">
         {# This part is shared between stand-alone WebView and Monochrome #}
diff --git a/android_webview/system_webview_apk_tmpl.gni b/android_webview/system_webview_apk_tmpl.gni
index ba0c2e6280ee7..68d76574b3c1c 100644
--- a/android_webview/system_webview_apk_tmpl.gni
+++ b/android_webview/system_webview_apk_tmpl.gni
@@ -99,6 +99,8 @@ template("system_webview_apk_or_module_tmpl") {
       } else {
         variables += [ "library=libmonochrome.so" ]
       }
+      _has_secondary_abi = !android_64bit_target_cpu || _include_32_bit_webview
+      variables += [ "has_secondary_abi=$_has_secondary_abi" ]
     }
     includes = []
     if (defined(invoker.jinja_input)) {
diff --git a/chrome/android/chrome_public_apk_tmpl.gni b/chrome/android/chrome_public_apk_tmpl.gni
index 8c04690cb1adf..485047f3395d6 100644
--- a/chrome/android/chrome_public_apk_tmpl.gni
+++ b/chrome/android/chrome_public_apk_tmpl.gni
@@ -208,6 +208,10 @@ template("chrome_common_apk_or_module_tmpl") {
     output = _android_manifest
     variables = default_chrome_public_jinja_variables +
                 [ "manifest_package=$_manifest_package" ]
+    if (_is_trichrome || _is_monochrome) {
+      _has_secondary_abi = !android_64bit_target_cpu || _include_32_bit_webview
+      variables += [ "has_secondary_abi=$_has_secondary_abi" ]
+    }
     if (_is_trichrome) {
       input = "//chrome/android/java/AndroidManifest_trichrome_chrome.xml"
       includes = [ "//chrome/android/java/AndroidManifest.xml" ]
diff --git a/chrome/android/java/AndroidManifest_monochrome.xml b/chrome/android/java/AndroidManifest_monochrome.xml
index f168f66b901ca..1fbbffc7ec95d 100644
--- a/chrome/android/java/AndroidManifest_monochrome.xml
+++ b/chrome/android/java/AndroidManifest_monochrome.xml
@@ -17,7 +17,11 @@
 
 {% block extra_application_attributes %}
 {{ super() }}
+{% if has_secondary_abi is defined and has_secondary_abi == 'false' %}
+android:multiArch="false"
+{% else %}
 android:multiArch="true"
+{% endif %}
 android:extractNativeLibs="false"
 {% if force_32_bit is defined and force_32_bit == 'true' %}
 android:use32bitAbi="true"
diff --git a/chrome/android/java/AndroidManifest_trichrome_chrome.xml b/chrome/android/java/AndroidManifest_trichrome_chrome.xml
index f4fb656373e3e..420048ff08a11 100644
--- a/chrome/android/java/AndroidManifest_trichrome_chrome.xml
+++ b/chrome/android/java/AndroidManifest_trichrome_chrome.xml
@@ -8,7 +8,11 @@
 
 {% block extra_application_attributes %}
 {{ super() }}
+{% if has_secondary_abi is defined and has_secondary_abi == 'false' %}
+android:multiArch="false"
+{% else %}
 android:multiArch="true"
+{% endif %}
 android:extractNativeLibs="false"
 {% endblock %}
 
